### DISCUSSION TITLE
Using multiple GPU

### DISCUSSION THREAD

**Discussion by edofazza** [talmolab/sleap#806] [Help!]
I have two GPUs and I want to use both for training, but the leap-train command takes as argument one GPU. Is there a way to use multiple GPUs at once to speed up the training?

**Comment by talmo** [comment]
Hi @edofazza,

This is not a currently available feature in SLEAP -- multi-GPU support is a little tricky and OS-dependent.

You're welcome to give it a go by following [this guide](https://www.tensorflow.org/guide/distributed_training) on TensorFlow multi-device strategies together with [this guide](https://sleap.ai/notebooks/Interactive_and_resumable_training.html) on training interactively with SLEAP.

You'll want comment out the `sleap.disable_preallocation()` so all GPUs are available and then something like:

```
trainer = ...  # build SLEAP trainer

mirrored_strategy = tf.distribute.MirroredStrategy()
with mirrored_strategy.scope():
    trainer.setup()
trainer.train()
```

You may need to tweak what goes inside the `with`, but give it a go and let us know if that works for you.

Cheers,

Talmo

**Comment by TheMalaland** [comment]
Hi @edofazza were you able to train with multiple gpu?

**Reply by edofazza** [reply]
Hi @TheMalaland, 

I haven't tried that yet since I got access to a different GPU with more memory, which it was enough for training my dataset.

**Reply by AmanRathoreP** [reply]
> Hi @edofazza were you able to train with multiple gpu?


you can try [sleap pytorch](https://github.com/talmolab/sleap-nn) it supports out of the box multiple GPU training. I'm attaching a sample config yaml file that consumes 4 GPU. I tested it with cuda 11.8 on 4 Tesla T4 GPUs
[snn_config.yaml](https://github.com/user-attachments/files/22067255/snn_config.yaml)

**Comment by AmanRathoreP** [comment]
check sleap's pytorch version it support [multi GPU training](https://nn.sleap.ai/latest/training/?h=gpu#multi-gpu-training).
